### ISSUE TITLE
(MapLink (ImplicationScope)) executes (GetLink) in consequent pattern body

### ISSUE THREAD

**Issue by vsbogd** [opencog/atomspace#2201]
I try to execute the following example:
```scm
(define (filter atom)
  (display "filter called ") (display atom) (newline)
  (SimpleTruthValue 1 1)
)

(cog-execute!
  (Map
    (ImplicationScope
      (TypedVariable (Variable "x") (Type "ConceptNode"))
      (Variable "x")
      (Get (And (Evaluation (GroundedPredicate "scm:filter") (Variable "x"))))
    )
    (Set (Concept "a") (Concept "b"))
  )
)
```
And I expect that it returns:
```scm
(Set
  (Get (And (Evaluation (GroundedPredicate "scm:filter") (Concept "a"))))
  (Get (And (Evaluation (GroundedPredicate "scm:filter") (Concept "b"))))
)
```
But instead it tries to execute `GetLink` without `Variable` substitution and throws exception:
```
filter called (VariableNode "x")

ERROR: In procedure opencog-extension:
Throw to key `C++-EXCEPTION' with args `("cog-execute!" " (/home/vital/projects/opencog/atomspace/opencog/query/Satisfier.cc:128)\nFunction args:\n((MapLink\n   (ImplicationScopeLink\n      (TypedVariableLink\n         (VariableNode \"x\")\n         (TypeNode \"ConceptNode\")\n      )\n      (VariableNode \"x\")\n      (GetLink\n         (AndLink\n            (EvaluationLink (stv 1 1)\n               (GroundedPredicateNode \"scm:filter\")\n               (VariableNode \"x\")\n            )\n         )\n      )\n   )\n   (SetLink\n      (ConceptNode \"b\")\n      (ConceptNode \"a\")\n   )\n)\n)")'.

Entering a new prompt.  Type `,bt' for a backtrace or `,q' to continue.
```

**Comment by vsbogd** [comment]
Depends on #2203

**Comment by linas** [comment]
Closing. This behavior stems from an old design decision that `cog-execute!` should run "all the way through" to the end, instead of single-stepping through some Atomese expression. It's not clear whether this could be or should be changed. I beleive that there's a large number of unit tests that expect execution to go "all the way"; each of these cases were designed to work that way for some reason. If there was some large team that could go back and review this design decision, it might be possible to change it. Alternatives, below.

First, the work-arounds: you can say either `(DontExec (Get ...))` or `(LocalQuote (Get ...))` and that will work (and it would have worked when this bug report was first filed; this is not new.)

You could use some non-executable link type, instead of `Get` -- e.g. `LambdaLink`

Finally, ***if*** `cog-execute!` was redesigned to single-step, then we would need to invent an `(ExecLink foo)` which would mean "when you see this, then do execute `foo`, because we want to take one more step". 

I have seen places where I said to myself, "gee, it would be nice to run this before that", and the `ExecLink` would allow that.  For example, inside a pattern, one might want to run the stuff in there first, to get the pattern, and only then perform the pattern match. 

Figuring out the "best" default control flow in Atomese remains tricky and confusing. Reviewing this would be a multi-month-long multi-person project.  Actually changing things around could take even longer.   Meanwhile, `DontExecLink` is your friend.

Closing as "wontfix"
